source/Mlos.Model.Services.Client/OptimizerServiceEncoderDecoder.cs

```diff
@@ -0,0 +1,419 @@
+            {
+                // A context space was provided
+                instance.ContextSpace = DecodeHypergrid(problem.ContextSpace);
+            }
+
+            return instance;
+        }
+
+        public OptimizationObjective DecodeOptimizationObjective(OptimizerService.Objective objective)
+        {
+            return new OptimizationObjective(objective.Name, objective.Minimize);
+        }
+
+        public EmptyDimension DecodeEmptyDimension(OptimizerService.EmptyDimension dimension)
+        {
+            return new EmptyDimension(dimension.Name, GrpcTypeToDimension(dimension.DimensionType));
+        }
+
+        public DiscreteDimension DecodeDiscreteDimension(OptimizerService.DiscreteDimension dimension)
+        {
+            return new DiscreteDimension(dimension.Name, dimension.Min, dimension.Max);
+        }
+
+        public CategoricalDimension DecodeCategoricalDimension(OptimizerService.CategoricalDimension dimension)
+        {
+            var values = new object[dimension.Values.Count];
+            var i = 0;
+            foreach (var value in dimension.Values)
+            {
+                values[i++] = DecodePrimitiveValue(value);
+            }
+
+            return new CategoricalDimension(dimension.Name, values);
+        }
+
+        public IDimension DecodeDimension(OptimizerService.Dimension dimension)
+        {
+            switch (dimension.DimensionCase)
+            {
+                case OptimizerService.Dimension.DimensionOneofCase.ContinuousDimension:
+                    return DecodeContinuousDimension(dimension.ContinuousDimension);
+                case OptimizerService.Dimension.DimensionOneofCase.DiscreteDimension:
+                    return DecodeDiscreteDimension(dimension.DiscreteDimension);
+                case OptimizerService.Dimension.DimensionOneofCase.OrdinalDimension:
+                    return DecodeOrdinalDimension(dimension.OrdinalDimension);
+                case OptimizerService.Dimension.DimensionOneofCase.CategoricalDimension:
+                    return DecodeCategoricalDimension(dimension.CategoricalDimension);
+                case OptimizerService.Dimension.DimensionOneofCase.CompositeDimension:
+                    return DecodeCompositeDimension(dimension.CompositeDimension);
+                case OptimizerService.Dimension.DimensionOneofCase.EmptyDimension:
+                    return DecodeEmptyDimension(dimension.EmptyDimension);
+                default:
+                    System.Console.Error.WriteLine("Invalid type provided to OptimizerServiceEncoderDecoder.Decoder.DecodeDimension(..). Type: " + dimension.GetType());
+
+                    return null;
+            }
+        }
+
+        public CompositeDimension DecodeCompositeDimension(OptimizerService.CompositeDimension dimension)
+        {
+            var chunks = new IDimension[dimension.Chunks.Count];
+            var i = 0;
+            foreach (var serialized_chunk in dimension.Chunks)

```

Comment:
https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/capitalization-conventions

Reply:
I think I fixed this particular instance of python style capitalization when I switched to Linq. 

[zaporter/grpc dd30df8] Updated to use Linq in the EncoderDecoder. Also changed Smartcache to use a primitive value rather than enum.